bindings/rust/src/pippenger-no_std.rs

```diff
@@ -49,13 +49,13 @@ macro_rules! pippenger_mult_impl {
             pub fn from(points: &[$point]) -> Self {
                 let npoints = points.len();
                 let mut ret = Self {
-                    points: Vec::with_capacity(npoints),
+                    points: vec![Default::default(); npoints],
```

Comment:
The intent here is to avoid a useless zero initialization when ret.points will be initialized 5 lines below.

Reply:
This is one of the examples of things that may work by accident, but absolutely not obvious to readers of the code or those who will audit code later whether this is safe or not, while new version is obviously safe.

If code below initializes input then this should use something like https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html instead of using brute force and resizing vector to cover uninitialized memory (that has potential of causing UB if read before writing).

Also this is precisely the kind of thing that compilers should be good at optimizing. I didn't check, but wouldn't be surprised if it makes zero difference in assembly.

Reply:
The compiler cannot optimize this away because the next call is through C FFI which is opaque.

Furthermore, the points may take GBs of memory as proof systems flirt with 2^26 points these days (67M points, each taking 96B in affine coordinates to 192 bytes in XYZZ extended jacobian coordinates).

Reply:
I believe it should be able to do so easily with LTO (here is an article from years ago: https://blog.llvm.org/2019/09/closing-gap-cross-language-lto-between.html)? And allocating zeroed memory is very fast from what I remember testing (at least on Linux), so if you really have gigabytes of memory to allocate, this line will be one of the last performance issues to resolve.

Reply:
There are several pitfalls with LTO:

1. Compilers are increasingly smart and to guarantee constant-time you either need assembly or (separate compilation units and no LTO), see https://www.cl.cam.ac.uk/~rja14/Papers/whatyouc.pdf and more generally my "Fighting the compiler" paragraph at https://github.com/mratsim/constantine/wiki/Constant-time-arithmetics#fighting-the-compiler
2. AFAIK each project will need a `.cargo/config.toml` with `rustflags="-Clinker-plugin-lto -Clinker=clang -Clink-arg=-fuse-ld=lld"`, requiring Clang + LLD installation so README/install has more friction.
3. This project does not use inline assembly but it's worth noting that Clang+inline assembly is broken when dealing with constants that can be propagated through inline assembly, see full details here: https://github.com/mratsim/constantine/pull/309#issuecomment-1837216115. Many projects use libsecp256k1 and blst for example.
4. Rust and Clang LLVM versions have to be the same, it may happen in some distros or in CI that they support a different IR, for example the transition to LLVM 15 changed the default pointer types from typed to opaque (or the inverse) and Rust and Clang in Github Actions were compiled from different LLVM version.

Due to this, I think LTO is significantly easier to deal with at application-level that have the luxury to ship a docker rather than libraries. But if we can get the benefit by just skipping zero initialization in a library, I think it's worth it.

That said, I agree on benchmarking to guide such decisions.